aip/client-libraries/4222.md

```diff
@@ -47,6 +47,11 @@ md := metadata.Pairs("x-goog-request-params",
 Much like URL parameters, if there is more than one key-value pair, the `&`
 character is used as the separator.
 
+If the [`google.api.http`][http] annotation contains `additional_bindings`,
```

Comment:
Do we need to explicitly state not to include parameters that the user does not provide?

Reply:
Hm...do you mean "don't include OUTPUT_ONLY fields"?

Reply:
No, I mean, let's say you have the following goofy binding...

```proto
rpc BurnBooks(BurnBooksRequest) returns (BurnBooksResponse) {
  option (google.api.http) = {
    post: "/v1/{publisher=publishers/*}/books:burn"
    body: "*"
    additional_bindings: {
      post: "/v1/{shelf=shelves/*}/books:burn"
      body: "*"
    }
  };
}

message BurnBooksRequest {
  oneof parent {  // Never do this; the field should have been "parent".
    string book = 1;
    string shelf = 2;
  }
}
```

The header should only include the things the user _sets_. So, in this case, the header we send should always be one of these two:

    publisher=publishers/foo
    shelf=shelves/foo

And _not_:

    publisher=publishers/foo&shelf=
    publisher=&shelf=shelves/foo

Reply:
I think I understand what you are suggesting. I think the better approach would be to discourage the use of oneofs in HTTP bindings? Is that fair to say?

Maybe the fields that are specified in the HTTP binding should only be those that are REQUIRED? In my limited experience, most of the HTTP variables are required variables, e.g. a resource name.

I'm hesitant to make this check explicit b.c it would add a lot of boiler plate code the clients, checking if every specified field is non-empty before adding it.

Reply:
It seems like that would abrogate the need for this change at all? If every HTTP binding can only include required fields, then they would all be present in every binding, yes?

Reply:
Ah you are right, not sure what I was thinking there.

The first suggestion still stands, to recommend not using oneof fields in HTTP binding variables. As does my concern about boiler plate.

Reply:
Yes, both of those are reasonable. But I believe we still need something that says "if a field empty, do not include it in the header".

Reply:
Done.